File: src/main/java/edu/cmu/ml/rtw/pra/graphs/GraphCreator.java
package edu.cmu.ml.rtw.pra.graphs;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.log4j.Logger;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Sets;

import edu.cmu.graphchi.ChiFilenames;
import edu.cmu.graphchi.EmptyType;
import edu.cmu.graphchi.datablocks.IntConverter;
import edu.cmu.graphchi.preprocessing.EdgeProcessor;
import edu.cmu.graphchi.preprocessing.FastSharder;
import edu.cmu.ml.rtw.users.matt.util.Dictionary;
import edu.cmu.ml.rtw.users.matt.util.FileUtil;
import edu.cmu.ml.rtw.users.matt.util.IntTriple;
import edu.cmu.ml.rtw.users.matt.util.MapUtil;
import edu.cmu.ml.rtw.users.matt.util.Pair;

/**
 * Given an (already constructed) KB graph, convert it to a GraphChi graph suitable for using with
 * the PRA code, including the necessary dictionary files.  This just produces three files: an
 * edges.tsv file that will be directly input to GraphChi, and two dictionaries, a node dictionary
 * and an edge dictionary, that give the mapping from the integers in the edges.tsv file to node
 * and edge names.
 */
public class GraphCreator {
  @SuppressWarnings("unused")
  private static Logger log = Logger.getLogger(GraphCreator.class);

  private final List<RelationSet> relationSets;
  private final String outdir;
  private final FileUtil fileUtil;
  private final boolean deduplicateEdges;
  private final boolean createMatrices;
  private final int maxMatrixFileSize;

  public GraphCreator(GraphConfig config) {
    this(config, new FileUtil());
  }

  @VisibleForTesting
  protected GraphCreator(GraphConfig config, FileUtil fileUtil) {
    this.relationSets = config.relationSets;
    if (!config.outdir.endsWith("/")) {
      this.outdir = config.outdir + "/";
    } else {
      this.outdir = config.outdir;
    }
    this.deduplicateEdges = config.deduplicateEdges;
    this.createMatrices = config.createMatrices;
    this.maxMatrixFileSize = config.maxMatrixFileSize;
    this.fileUtil = fileUtil;
  }

  public void createGraphChiRelationGraph() throws IOException {
    createGraphChiRelationGraph(true);
  }

  public void createGraphChiRelationGraph(boolean shardGraph) throws IOException {
    // Some preparatory stuff
    fileUtil.mkdirOrDie(outdir);

    fileUtil.mkdirs(outdir + "graph_chi/");
    String edgeFilename = outdir + "graph_chi/edges.tsv";
    FileWriter intEdgeFile = fileUtil.getFileWriter(edgeFilename);

    List<Pair<String, Map<String, List<String>>>> aliases = Lists.newArrayList();
    for (RelationSet relationSet : relationSets) {
      if (relationSet.getIsKb()) {
        String aliasRelation = relationSet.getAliasRelation();
        Map<String, List<String>> kbAliases = relationSet.getAliases();
        aliases.add(Pair.makePair(aliasRelation, kbAliases));
      }
    }

    Dictionary nodeDict = new Dictionary();
    Dictionary edgeDict = new Dictionary();

    Set<String> seenNps = Sets.newHashSet();
    Set<IntTriple> seenTriples = null;
    if (deduplicateEdges) {
      seenTriples = Sets.newHashSet();
    }
    Map<RelationSet, String> prefixes = getSvoPrefixes();
    int numEdges = 0;
    for (RelationSet relationSet : relationSets) {
      System.out.println("Adding edges to the graph from " + relationSet.getRelationFile());
      String prefix = null;
      if (prefixes != null) {
        prefix = prefixes.get(relationSet);
      }
      numEdges += relationSet.writeRelationEdgesToGraphFile(intEdgeFile,
                                                            seenTriples,
                                                            prefix,
                                                            seenNps,
                                                            aliases,
                                                            nodeDict,
                                                            edgeDict);
    }
    intEdgeFile.close();

    // Adding edges is now finished, and the dictionaries aren't getting any more entries, so we
    // can output them.
    outputDictionariesToDisk(nodeDict, edgeDict);

    // Now decide how many shards to do, based on the number of edges that are in the graph.
    int numShards = getNumShards(numEdges);
    FileWriter writer = fileUtil.getFileWriter(outdir + "num_shards.tsv");
    writer.write(numShards + "\n");
    writer.close();
    if (shardGraph) {
      shardGraph(edgeFilename, numShards);
    }

    // This is for if you want to do the path following step with matrix multiplications instead of
    // with random walks (which I'm expecting to be a lot faster, but haven't finished implementing
    // yet).
    if (createMatrices) {
      outputMatrices(fileUtil.getBufferedReader(edgeFilename), edgeDict.getNextIndex());
    }
  }

  /**
   * Runs GraphChi's preprocessing (sharding) on the graph.  This produces a number of shard files,
   * and if the files are already present, this is a no-op.  So it's only run once for each graph,
   * no matter how many times you run GraphChi code.
   */
  private void shardGraph(String baseFilename, int numShards) throws IOException {
    FastSharder<EmptyType, Integer> sharder = new FastSharder<EmptyType, Integer>(baseFilename, numShards, null,
        new EdgeProcessor<Integer>() {
          public Integer receiveEdge(int from, int to, String token) {
            return Integer.parseInt(token);
          }
        }, null, new IntConverter());
    if (!new File(ChiFilenames.getFilenameIntervals(baseFilename, numShards)).exists()) {
      sharder.shard(new FileInputStream(new File(baseFilename)), "edgelist");
    }
  }

  ////////////////////////////////////////////////////////
  // Other boilerplate
  ////////////////////////////////////////////////////////

  public void outputDictionariesToDisk(Dictionary nodeDict,
                                       Dictionary edgeDict) throws IOException {
    System.out.println("Outputting dictionaries to disk");
    FileWriter nodeDictFile = fileUtil.getFileWriter(outdir + "node_dict.tsv");
    nodeDict.writeToWriter(nodeDictFile);
    nodeDictFile.close();

    FileWriter edgeDictFile = fileUtil.getFileWriter(outdir + "edge_dict.tsv");
    edgeDict.writeToWriter(edgeDictFile);
    edgeDictFile.close();
  }

  public int getNumShards(int numEdges) {
    if (numEdges < 5000000) {
      return 2;
    } else if (numEdges < 10000000) {
      return 3;
    } else if (numEdges < 40000000) {
      return 4;
    } else if (numEdges < 100000000) {
      return 5;
    } else if (numEdges < 150000000) {
      return 6;
    } else if (numEdges < 250000000) {
      return 7;
    } else if (numEdges < 350000000) {
      return 8;
    } else if (numEdges < 500000000) {
      return 9;
    } else {
      return 10;
    }
  }

  @VisibleForTesting
  protected void outputMatrices(BufferedReader reader, int numRelations) throws IOException {
    System.out.println("Creating matrices");
    fileUtil.mkdirs(outdir + "matrices/");
    Map<Integer, List<Pair<Integer, Integer>>> matrices = Maps.newHashMap();
    System.out.println("Reading edge file");
    String line;
    while ((line = reader.readLine()) != null) {
      String fields[] = line.split("\t");
      int source = Integer.parseInt(fields[0]);
      int target = Integer.parseInt(fields[1]);
      int relation = Integer.parseInt(fields[2]);
      MapUtil.addValueToKeyList(matrices, relation, Pair.makePair(source, target));
    }
    reader.close();
    System.out.println("Outputting matrix files");
    List<List<Pair<Integer, Integer>>> edgesToWrite = Lists.newArrayList();
    int startRelation = 1;
    int edgesSoFar = 0;
    for (int i = 1; i <= numRelations; i++) {
      List<Pair<Integer, Integer>> instances = matrices.get(i);
      if (edgesSoFar > 0 && edgesSoFar + instances.size() > maxMatrixFileSize) {
        writeEdgesSoFar(startRelation, i - 1, edgesToWrite);
        startRelation = i;
      }
      edgesToWrite.add(instances);
      edgesSoFar += instances.size();
    }
    if (edgesToWrite.size() > 0) {
      writeEdgesSoFar(startRelation, numRelations, edgesToWrite);
    }
    System.out.println("Done creating matrices");
  }

  private void writeEdgesSoFar(int startRelation,
                               int endRelation,
                               List<List<Pair<Integer, Integer>>> edgesToWrite) throws IOException {
    String filename = outdir + "matrices/" + startRelation;
    if (endRelation > startRelation) {
      filename += "-" + endRelation;
    }
    FileWriter writer = fileUtil.getFileWriter(filename);
    for (List<Pair<Integer, Integer>> relation_instances : edgesToWrite) {
      writer.write("Relation " + startRelation + "\n");
      for (Pair<Integer, Integer> instance : relation_instances) {
        writer.write(instance.getLeft() + "\t" + instance.getRight() + "\n");
      }
      startRelation++;
    }
    writer.close();
    edgesToWrite.clear();
  }

  /**
   * Create a prefix for each SVO file as necessary, according to how they were embedded.
   *
   * If the edges are embedded, we need to differentiate the latent representations if they were
   * not made together.  That is, if we have two or more embedded SVO files, and they have
   * embeddings files that are _different_, that means that a +L1 edge from one and a +L1 edge from
   * another are not the same edge type.  So we add a prefix to the edge type that is specific to
   * each embedding.  This isn't a problem with KB edges vs. SVO edges, because the "alias"
   * relation assures that the two kinds of edges will never share the same space.
   */
  public Map<RelationSet, String> getSvoPrefixes() {
    Map<String, List<RelationSet>> embeddingsToRels = Maps.newHashMap();
    for (RelationSet relationSet : relationSets) {
      if (relationSet.getIsKb()) continue;
      if (relationSet.getEmbeddingsFile() == null) continue;
      MapUtil.addValueToKeyList(embeddingsToRels, relationSet.getEmbeddingsFile(), relationSet);
    }
    Map<RelationSet, String> prefixes = null;
    if (embeddingsToRels.size() > 1) {
      prefixes = Maps.newHashMap();
      int prefix = 0;
      for (Map.Entry<String, List<RelationSet>> entry : embeddingsToRels.entrySet()) {
        prefix++;
        for (RelationSet relationSet : entry.getValue()) {
          prefixes.put(relationSet, prefix + "-");
        }
      }
    }
    return prefixes;
  }
}

File: src/main/scala/edu/cmu/ml/rtw/pra/graphs/GraphCreator.scala
package edu.cmu.ml.rtw.pra.graphs

import java.io.BufferedReader
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter

import edu.cmu.graphchi.ChiFilenames
import edu.cmu.graphchi.EmptyType
import edu.cmu.graphchi.datablocks.IntConverter
import edu.cmu.graphchi.preprocessing.EdgeProcessor
import edu.cmu.graphchi.preprocessing.FastSharder
import edu.cmu.ml.rtw.pra.experiments.Outputter

import com.mattg.pipeline.Step
import com.mattg.util.Dictionary
import com.mattg.util.FileUtil
import com.mattg.util.IntTriple
import com.mattg.util.JsonHelper
import com.mattg.util.MutableConcurrentDictionary
import com.mattg.util.Pair

import scala.collection.JavaConverters._
import scala.collection.mutable

import org.json4s._
import org.json4s.JsonDSL.WithDouble._
import org.json4s.native.JsonMethods._

// TODO(matt): Add an option to exclude the test edges in a split when creating the graph.  This
// makes the graph dependent on the split, but is necessary for some experimental protocols.  It
// should be easy to use seenTriples to implement this - just add the whole test split (including
// inverses) to the seenTriples object before adding any edges.
class GraphCreator(
  baseGraphDir: String,
  params: JValue,
  outputter: Outputter,
  fileUtil: FileUtil
) extends Step(Some(params), fileUtil) {
  implicit val formats = DefaultFormats
  override val name = "Graph Creator"

  val graphName = (params \ "name").extract[String]
  val outdir = s"$baseGraphDir/$graphName/"
  override val inProgressFile = outdir + "in_progress"
  override val paramFile = outdir + "params.json"

  val relationSets = (params \ "relation sets").children.map(relationSet => {
    (relationSet \ "type") match {
      case JNothing => new RelationSet(relationSet, outputter, fileUtil)
      case JString("generated") => generateSyntheticRelationSet(relationSet \ "generation params")
      case other => throw new IllegalStateException("Bad relation set specification")
    }
  })

  // TODO(matt): this could use some work.  Really, we just want to take the union of the required
  // inputs to the RelationSets.  And we want to use None instead of null here.  But, that'll wait
  // for another day.
  val inputRelationFiles = relationSets.map(_.relationFile)
  val inputAliasFiles = relationSets.map(_.aliasFile).filterNot(_ == null)
  override val inputs: Set[(String, Option[Step])] = (inputRelationFiles ++ inputAliasFiles).map((_, None)).toSet

  val deduplicateEdges = JsonHelper.extractWithDefault(params, "deduplicate edges", false)
  val outputBinaryFile = JsonHelper.extractWithDefault(params, "output binary file", true)
  val outputPlainTextFile = JsonHelper.extractWithDefault(params, "output plain text file", false)
  val shouldShardGraph = JsonHelper.extractWithDefault(params, "shard plain text graph", false)
  val outputMatrices = JsonHelper.extractWithDefault(params, "output matrices", false)
  val maxMatrixFileSize = JsonHelper.extractWithDefault(params, "max matrix file size", 100000)

  val matrixOutDir = outdir + "matrices/"
  val plainTextFilename = outdir + "graph_chi/edges.tsv"
  val binaryFilename = outdir + "edges.dat"
  val nodeDictFilename = outdir + "node_dict.tsv"
  val edgeDictFilename = outdir + "edge_dict.tsv"

  override val outputs = {
    val tmp = new mutable.ListBuffer[String]
    tmp += outdir
    tmp += nodeDictFilename
    tmp += edgeDictFilename
    if (outputBinaryFile) tmp += binaryFilename
    if (outputPlainTextFile) tmp += plainTextFilename
    if (outputMatrices) tmp += matrixOutDir
    tmp.toSet
  }

  // This is from old, experimental code that didn't pan out...
  override def parametersMatch(params1: JValue, params2: JValue): Boolean = {
    super.parametersMatch(
      params1.removeField(_._1.equals("denser matrices")),
      params2.removeField(_._1.equals("denser matrices"))
    )
  }

  override def _runStep() {
    outputter.info(s"Creating graph $name in $outdir")

    outputter.info("Loading aliases")
    val aliases = relationSets.filter(_.isKb).par.map(relationSet => {
      (relationSet.aliasRelation, relationSet.getAliases)
    }).seq

    val nodeDict = new MutableConcurrentDictionary()
    val edgeDict = new MutableConcurrentDictionary()

    createGraphFiles(aliases, nodeDict, edgeDict)

    // Adding edges is now finished, and the dictionaries aren't getting any more entries, so we
    // can output them.
    outputter.info("Outputting dictionaries to disk")
    val nodeDictFile = fileUtil.getFileWriter(nodeDictFilename)
    val edgeDictFile = fileUtil.getFileWriter(edgeDictFilename)
    nodeDict.writeToWriter(nodeDictFile)
    nodeDictFile.close()
    edgeDict.writeToWriter(edgeDictFile)
    edgeDictFile.close()

    // This is for if you want to do the path following step with matrix multiplications instead of
    // with random walks.  This was basically a failed experiment, and is not recommended.
    if (outputMatrices) {
      createMatrices(outdir + "graph_chi/edges.tsv", maxMatrixFileSize)
    }
  }

  def createGraphFiles(
    aliases: Seq[(String, Map[String, Seq[String]])],
    nodeDict: MutableConcurrentDictionary,
    edgeDict: MutableConcurrentDictionary
  ) {
    val seenNps = new mutable.HashSet[String]
    val seenTriples: mutable.HashSet[(Int, Int, Int)] = {
      if (deduplicateEdges) {
        new mutable.HashSet[(Int, Int, Int)]
      } else {
        null
      }
    }
    val prefixes = getSvoPrefixes(relationSets)

    val plainTextFile = if (outputPlainTextFile) {
      fileUtil.mkdirs(outdir + "graph_chi/")
      Some(fileUtil.getFileWriter(plainTextFilename))
    } else {
      None
    }

    val binaryFile = if (outputBinaryFile) {
      Some(fileUtil.getDataOutputStream(outdir + "edges.dat"))
    } else {
      None
    }

    var numEdges = 0
    for (relationSet <- relationSets) {
      outputter.info("Adding edges to the graph from " + relationSet.relationFile)
      val prefix = prefixes(relationSet)
      numEdges += relationSet.writeRelationEdgesToGraphFile(
        plainTextFile,
        binaryFile,
        seenTriples,
        prefix,
        seenNps,
        aliases,
        nodeDict,
        edgeDict
      )
    }
    plainTextFile.foreach(_.close())
    binaryFile.foreach(_.close())

    if (outputPlainTextFile && shouldShardGraph) {
      // Now decide how many shards to do, based on the number of edges that are in the graph.
      val numShards = getNumShards(numEdges)
      val writer = fileUtil.getFileWriter(outdir + "num_shards.tsv")
      writer.write(numShards + "\n")
      writer.close()
      if (shouldShardGraph) {
        shardGraph(plainTextFilename, numShards)
      }
    }
  }

  /**
   * Runs GraphChi's preprocessing (sharding) on the graph.  This produces a number of shard files,
   * and if the files are already present, this is a no-op.  So it's only run once for each graph,
   * no matter how many times you run GraphChi code.
   */
  def shardGraph(baseFilename: String, numShards: Int) {
    val sharder = new FastSharder[EmptyType, Integer](baseFilename, numShards, null,
        new EdgeProcessor[Integer]() {
          def receiveEdge(from: Int, to: Int, token: String): java.lang.Integer = {
            token.toInt
          }
        }, null, new IntConverter())
    if (!new File(ChiFilenames.getFilenameIntervals(baseFilename, numShards)).exists()) {
      sharder.shard(new FileInputStream(new File(baseFilename)), "edgelist")
    }
  }

  ////////////////////////////////////////////////////////
  // Other boilerplate
  ////////////////////////////////////////////////////////

  def getNumShards(numEdges: Int) = {
    if (numEdges < 5000000) {
      2
    } else if (numEdges < 10000000) {
      3
    } else if (numEdges < 40000000) {
      4
    } else if (numEdges < 100000000) {
      5
    } else if (numEdges < 150000000) {
      6
    } else if (numEdges < 250000000) {
      7
    } else if (numEdges < 350000000) {
      8
    } else if (numEdges < 500000000) {
      9
    } else {
      10
    }
  }

  def createMatrices(filename: String, maxMatrixFileSize: Int) {
    outputter.info("Creating matrices")
    fileUtil.mkdirs(outdir + "matrices/")
    outputter.info("Reading edge file")
    var line: String = null
    val lines = fileUtil.readLinesFromFile(filename)
    val matrices = lines.par.map(line => {
      val fields = line.split("\t")
      (fields(0).toInt, fields(1).toInt, fields(2).toInt)
    }).groupBy(x => x._3).toMap.mapValues(triple_set => {
      triple_set.map(triple => (triple._1, triple._2)).seq.toSeq
    }).seq
    val numRelations = matrices.map(_._1).max
    outputter.info("Outputting matrix files")
    val edgesToWrite = new mutable.ArrayBuffer[Seq[(Int, Int)]]
    var startRelation = 1
    var edgesSoFar = 0
    for (i <- 1 to numRelations) {
      val matrix = matrices.getOrElse(i, Nil)
      if (matrix.size == 0) outputter.warn("RELATION WITH NO INSTANCES: " + i)
      if (edgesSoFar > 0 && edgesSoFar + matrix.size > maxMatrixFileSize) {
        writeEdgesSoFar(startRelation, i - 1, edgesToWrite.toSeq)
        edgesToWrite.clear
        startRelation = i
        edgesSoFar = 0
      }
      edgesToWrite += matrix
      edgesSoFar += matrix.size
    }
    if (edgesToWrite.size > 0) {
      writeEdgesSoFar(startRelation, numRelations, edgesToWrite)
    }
    outputter.info("Done creating matrices")
  }

  def writeEdgesSoFar(_start_relation: Int, end_relation: Int, edges_to_write: Seq[Seq[(Int, Int)]]) {
    var start_relation = _start_relation
    var filename = matrixOutDir + start_relation
    if (end_relation > start_relation) {
      filename += "-" + end_relation
    }
    val writer = fileUtil.getFileWriter(filename)
    for (matrix <- edges_to_write) {
      writer.write("Relation " + start_relation + "\n")
      for (entry <- matrix) {
        writer.write(entry._1 + "\t" + entry._2 + "\n")
      }
      start_relation += 1
    }
    writer.close()
  }

  /**
   * Create a prefix for each SVO file as necessary, according to how they were embedded.
   *
   * If the edges are embedded, we need to differentiate the latent representations if they were
   * not made together.  That is, if we have two or more embedded SVO files, and they have
   * embeddings files that are _different_, that means that a +L1 edge from one and a +L1 edge from
   * another are not the same edge type.  So we add a prefix to the edge type that is specific to
   * each embedding.  This isn't a problem with KB edges vs. SVO edges, because the "alias"
   * relation assures that the two kinds of edges will never share the same space.
   */
  def getSvoPrefixes(relationSets: Seq[RelationSet]): Map[RelationSet, String] = {
    val embeddingsToRels = relationSets
      .filter(!_.isKb)
      .filter(_.embeddingsFile != null).map(relationSet => {
        (relationSet.embeddingsFile, relationSet)
      }).groupBy(_._1).toMap
    if (embeddingsToRels.size <= 1) {
      Map[RelationSet, String]().withDefaultValue(null)
    } else {
      embeddingsToRels.zipWithIndex.flatMap(x => {
        x._1._2.map(y => (y._2, s"${x._2 + 1}-"))
      })
    }
  }

  var synthetic_data_creator_factory: ISyntheticDataCreatorFactory = new SyntheticDataCreatorFactory

  def generateSyntheticRelationSet(params: JValue): RelationSet = {
    val baseDir = baseGraphDir.replace("graphs/", "")
    val creator = synthetic_data_creator_factory.getSyntheticDataCreator(baseDir, params, outputter, fileUtil)
    if (fileUtil.fileExists(creator.relation_set_dir)) {
      fileUtil.blockOnFileDeletion(creator.in_progress_file)
      val current_params = parse(fileUtil.readLinesFromFile(creator.param_file).mkString("\n"))
      if (current_params.equals(JNothing)) {
        outputter.warn(s"Odd...  couldn't read parameters from ${creator.param_file}, even though " +
          s"${creator.relation_set_dir} exists")
      }
      if (current_params != params) {
        outputter.fatal(s"Parameters found in ${creator.param_file}: ${pretty(render(current_params))}")
        outputter.fatal(s"Parameters specified in spec file: ${pretty(render(params))}")
        outputter.fatal(s"Difference: ${current_params.diff(params)}")
        throw new IllegalStateException("Synthetic data parameters don't match!")
      }
    } else {
      creator.createRelationSet()
    }
    val relSetParams = ("relation file" -> creator.data_file) ~ ("is kb" -> false)
    new RelationSet(relSetParams, outputter, fileUtil)
  }
}
